client/protocols/export.py

```diff
@@ -0,0 +1,145 @@
+from client.utils import network
+from datetime import datetime
+import client
+import logging
+import os
+import pickle
+import sys
+
+log = logging.getLogger(__name__)
+
+GAE_DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S.%f"
+
+class ExportProtocol(models.Protocol):
+    """Downloads submissions from the server"""
+
+    def on_interact(self):
+        """If the --export parameter was used, downloads all submissions 
+        for the given assignment from the server and then exits
+        """
+        data = None
+        try:
+            data = pickle.load(open("export_cache.pkl", "rb"))
+            self.access_token = auth.authenticate(self.args.authenticate)
+            log.info('Authenticated with access token %s', self.access_token)
+        except Exception as e:
+            if self.args.export:
+                data = {}
+                self.access_token = auth.authenticate(self.args.authenticate)
+                log.info('Authenticated with access token %s', self.access_token)
+                data['assign'] = self.args.export
+                data['course'] = self.get_course()
+                data['students'] = self.get_students(data['course'])
+                data['assignid'] = self.get_assign_id(
+                    data['course'], data['assign'])
+                data['current'] = 0
+        
+        if data:
+            try:
+                if not os.path.exists(data['assign']): 
+                    os.makedirs(data['assign'])
+                
+                downloads_left = len(data['students']) - data['current']
+                
+                print("{0} submissions left to download".format(downloads_left))
+                
+                for i in range(data['current'], len(data['students'])):
+                    student = data['students'][i]
+                    try:
+                        self.download_submission(student, data['assign'], 
+                            data['assignid'])
+                        data['current'] = i + 1
+                    except Exception as e:
+                        data['current'] = i
+                        pickle.dump(data, open("export_cache.pkl", "wb"))
+                        dl_left = len(data['students']) - i
+                        print("Download failed. Run command again to continue.")
+                        print("{0} submissions left to download".format(dl_left))
+                        sys.exit(0)
+                        
+            except KeyboardInterrupt:
+                pickle.dump(data, open("export_cache.pkl", "wb"))
+                dl_left = len(data['students']) - i
+                print("Download interrupted. Run command again to continue.")
+                print("{0} submissions left to download".format(dl_left))
+                sys.exit(0)
+                
+            print("Submissions downloaded.")
+            if os.path.exists('export_cache.pkl'):
+                os.remove('export_cache.pkl')
+            sys.exit(0)
```

Comment:
I think there will be a family of protocols that do something like this: make API call to the server to either fetch or update some state and then exit. I don't want to keep adding things to ok-main every time that happens since that'll get cluttered.

It might be good to then subclass things? like your GradingProcotol and models framework?